### PR TITLE
Dynamic Domain Generalization

### PR DESCRIPTION
**Pull request by siaimes**
Dynamic Domain Generalization: https://arxiv.org/abs/2205.13913

Signed-off-by: siaimes <34199488+siaimes@users.noreply.github.com>

### CODE REVIEW

File: dassl/modeling/backbone/__init__.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
     shufflenet_v2_x0_5, shufflenet_v2_x1_0, shufflenet_v2_x1_5,
     shufflenet_v2_x2_0
 )
+from .resnet_dynamic import *
```
Comment (KaiyangZhou) [inline dassl/modeling/backbone/__init__.py:25]: try to avoid using `*`

File: dassl/modeling/ops/conv.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+"""
```
Comment (KaiyangZhou) [inline dassl/modeling/ops/conv.py:1]: I suggest changing `conv.py` to `conv2d_dynamic.py`

**Comment by KaiyangZhou** [comment]
Seems like the changes are safe and won't affect other methods. Have you checked this carefully?

Just double check, have you run "linter.sh" to remove potential inconsistencies?

File: dassl/modeling/backbone/__init__.py (hunk 1 above)
Comment (siaimes) [reply dassl/modeling/backbone/__init__.py:25]: I have added `__all__` to the source file, so only the parts where `__all__` is declared will be imported.

https://github.com/KaiyangZhou/Dassl.pytorch/pull/43/files#diff-d7eabb476b680df417ee0042708ce5ec179cdb63cd37194c94481a338a35612fR17-R24

```
__all__ = [
    'resnet18_dynamic', 'resnet50_dynamic', 'resnet101_dynamic',
    'resnet18_dynamic_ms_l123', 'resnet18_dynamic_ms_l12',
    'resnet18_dynamic_ms_l1', 'resnet50_dynamic_ms_l123',
    'resnet50_dynamic_ms_l12', 'resnet50_dynamic_ms_l1',
    'resnet101_dynamic_ms_l123', 'resnet101_dynamic_ms_l12',
    'resnet101_dynamic_ms_l1'
]
```

Comment (siaimes) [reply dassl/modeling/backbone/__init__.py:25]: I think this scheme would be better? What do you think?

File: dassl/modeling/ops/conv.py (hunk 2 above)
Comment (siaimes) [reply dassl/modeling/ops/conv.py:1]: The advantage of naming it `conv.py` is that if there are new convolutional layers in the future, they can be placed in this file. It does not only contain Conv2dDynamic, although it is currently the case.

**Comment by siaimes** [comment]
> Seems like the changes are safe and won't affect other methods. Have you checked this carefully?
> 
> Just double check, have you run "linter.sh" to remove potential inconsistencies?

1. Yes;
2. Yes.

File: dassl/modeling/backbone/__init__.py (hunk 1 above)
Comment (KaiyangZhou) [reply dassl/modeling/backbone/__init__.py:25]: yes, looks better!

could you also change single quotes to double quotes so the style is consistent throughout the code? thanks

File: dassl/modeling/ops/conv.py (hunk 2 above)
Comment (KaiyangZhou) [reply dassl/modeling/ops/conv.py:1]: ok makes sense

**Comment by KaiyangZhou** [comment]
If there is no new in DomainnetDG, please remove it. `train_u` won't be used in this case https://github.com/KaiyangZhou/Dassl.pytorch/blob/master/dassl/engine/trainer.py#L593

It'd be good if you can provide some configs in `https://github.com/KaiyangZhou/Dassl.pytorch/tree/master/configs/trainers/dg`

**Comment by siaimes** [comment]
> It'd be good if you can provide some configs in `https://github.com/KaiyangZhou/Dassl.pytorch/tree/master/configs/trainers/dg`

In my opinion, it would be better for the config files to be independent of Dassl, which I have placed here: https://github.com/MetaVisionLab/DDG/tree/main/dassl, where I also provide the startup script.

**Pull request merged**